File: tools/update_ledger.py
#!/usr/bin/env python3
"""Regenerate ledger/sot.json (SHA-256) for SoT files.

Usage:
  python tools/update_ledger.py

Notes:
- Excludes ledger/sot.json and ledger/checksum.asc to avoid self-reference loops.
- Intended for local use; CI only verifies.
"""
from __future__ import annotations
import datetime
import hashlib, json, os
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]

INCLUDE_DIRS = ["core","system","governance","metrics","mind","appendix","canon","runtime","tools",".github","docs"]
INCLUDE_FILES = ["manifest.yml","README.md","CONTRIBUTING.md","ISKRA_MANIFEST.md","LIBER_INITIUM.md"]

EXCLUDE = {
    Path("ledger/sot.json"),
    Path("ledger/checksum.asc"),
}

CHECKSUM_DEFAULTS = {
    # Human-readable checksum header (kept stable for verifiers)
    "version": "vΩ.1.2",
    "revision": "rev13-maki-priority+integrity",
    "algorithm": "sha256",
}

def sha256_file(path: Path) -> str:
    h = hashlib.sha256()
    with path.open("rb") as f:
        for chunk in iter(lambda: f.read(1024*1024), b""):
            h.update(chunk)
    return h.hexdigest()

def should_exclude(rel_path: Path) -> bool:
    """Check if a path should be excluded from the ledger."""
    parts = rel_path.parts
    path_str = str(rel_path)
    
    # SECURITY: Exclude __pycache__ directories and .pyc files to prevent bytecode smuggling
    if "__pycache__" in parts or path_str.endswith(".pyc"):
        return True
    
    # Exclude node_modules anywhere in the path
    if "node_modules" in parts:
        return True
    
    # Exclude build artifacts
    # Exclude coverage reports
    if "coverage" in parts:
        return True

    if "dist" in parts:
        return True
    
    # Exclude TypeScript build info files
    if path_str.endswith(".tsbuildinfo"):
        return True
    
    return False

def main() -> None:
    out = {"version": "sot-ledger/1", "sha256": {}}
    skipped = []
    
    # directories
    for d in INCLUDE_DIRS:
        p = ROOT / d
        if not p.exists():
            continue
        for file in sorted(p.rglob("*")):
            if file.is_dir():
                continue
            rel = file.relative_to(ROOT)
            if rel in EXCLUDE:
                continue
            
            # Skip node_modules, dist, and build artifacts
            if should_exclude(rel):
                continue
            
            # Try to get a valid UTF-8 path string
            try:
                path_str = str(rel).replace(os.sep,"/")
                # Test if it can be encoded/decoded properly
                path_str.encode('utf-8').decode('utf-8')
                # Also test JSON serialization
                json.dumps(path_str)
                out["sha256"][path_str] = sha256_file(file)
            except (UnicodeError, UnicodeDecodeError, UnicodeEncodeError) as e:
                # Skip files with invalid UTF-8 filenames
                skipped.append((str(rel), str(e)))
                continue

    # top-level files
    for f in INCLUDE_FILES:
        p = ROOT / f
        if p.exists() and p.is_file():
            rel = p.relative_to(ROOT)
            if rel not in EXCLUDE:
                try:
                    path_str = str(rel).replace(os.sep,"/")
                    path_str.encode('utf-8').decode('utf-8')
                    out["sha256"][path_str] = sha256_file(p)
                except (UnicodeError, UnicodeDecodeError, UnicodeEncodeError) as e:
                    skipped.append((str(rel), str(e)))
                    continue

    ledger = ROOT / "ledger"
    ledger.mkdir(exist_ok=True)
    (ledger / "sot.json").write_text(json.dumps(out, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
    print(f"Updated {ledger/'sot.json'} with {len(out['sha256'])} entries")
    
    if skipped:
        print(f"\n⚠ Skipped {len(skipped)} files with encoding issues:")
        for path, error in skipped[:10]:  # Show first 10
            print(f"  - {path}")
        if len(skipped) > 10:
            print(f"  ... and {len(skipped) - 10} more")

    # Also keep a human-readable checksum file in sync.
    # This is NOT included in sot.json to avoid self-reference loops.
    checksum_path = ledger / "checksum.asc"
    meta = dict(CHECKSUM_DEFAULTS)
    if checksum_path.exists():
        # Preserve version/revision if they were manually bumped.
        for line in checksum_path.read_text(encoding="utf-8").splitlines():
            if line.startswith("version:"):
                meta["version"] = line.split(":", 1)[1].strip()
            elif line.startswith("revision:"):
                meta["revision"] = line.split(":", 1)[1].strip()
            elif line.startswith("algorithm:"):
                meta["algorithm"] = line.split(":", 1)[1].strip()

    meta["updated"] = datetime.date.today().isoformat()

    lines = [
        "-----BEGIN ISKRA CHECKSUM-----",
        f"version: {meta['version']}",
        f"revision: {meta['revision']}",
        f"updated: {meta['updated']}",
        f"algorithm: {meta['algorithm']}",
        "",
        "# path  sha256",
    ]
    for rel in sorted(out["sha256"].keys()):
        lines.append(f"{rel}  {out['sha256'][rel]}")
    lines.append("-----END ISKRA CHECKSUM-----")
    checksum_path.write_text("\n".join(lines) + "\n", encoding="utf-8")
    print(f"Updated {checksum_path}")

if __name__ == "__main__":
    main()

File: tools/update_ledger.ts
import * as fs from 'fs';
import * as path from 'path';
import * as crypto from 'crypto';
import { spawnSync } from 'child_process';

/**
 * TypeScript replacement for update_ledger.py
 * Maintains Iskra Source of Truth SHA-256 integrity.
 */

const ROOT = process.cwd();
const INCLUDE_DIRS = ["core","system","governance","metrics","mind","appendix","canon","runtime","tools",".github","docs","supabase"];
const INCLUDE_FILES = ["manifest.yml","README.md","CONTRIBUTING.md","ISKRA_MANIFEST.md","LIBER_INITIUM.md","pnpm-lock.yaml"];
const EXCLUDE = new Set(["ledger/sot.json", "ledger/checksum.asc"]);

const CHECKSUM_DEFAULTS = {
    version: "vΩ.1.2",
    revision: "rev13-maki-priority+integrity",
    algorithm: "sha256"
};

function sha256_file(filePath: string): string {
    // Normalize line endings to LF for consistent hashing across Windows/Linux
    const content = fs.readFileSync(filePath, 'utf-8').replace(/\r\n/g, '\n');
    const hashSum = crypto.createHash('sha256');
    hashSum.update(content, 'utf-8');
    return hashSum.digest('hex');
}

function should_exclude(rel_path: string): boolean {
    const parts = rel_path.split('/');
    if (parts.includes('__pycache__') || rel_path.endsWith('.pyc')) return true;
    if (parts.includes('node_modules')) return true;
    if (parts.includes('coverage')) return true;
    if (parts.includes('dist')) return true;
    if (rel_path.endsWith('.tsbuildinfo')) return true;
    return false;
}

function trackedFiles(): string[] {
    const pathspecs = [...INCLUDE_DIRS, ...INCLUDE_FILES];
    const result = spawnSync('git', ['ls-files', '-z', '--', ...pathspecs], {
        cwd: ROOT,
        encoding: 'utf8',
        windowsHide: true,
        maxBuffer: 16 * 1024 * 1024,
    });

    if (result.error) throw result.error;
    if (result.status !== 0) {
        throw new Error(`git ls-files failed with ${result.status}: ${result.stderr.trim()}`);
    }

    return result.stdout
        .split('\0')
        .map((item) => item.trim())
        .filter(Boolean)
        .sort();
}

const out = { version: "sot-ledger/1", sha256: {} as Record<string, string> };

for (const rel of trackedFiles()) {
    if (EXCLUDE.has(rel) || should_exclude(rel)) continue;
    const file = path.join(ROOT, rel);
    if (!fs.existsSync(file) || !fs.lstatSync(file).isFile()) continue;
    out.sha256[rel] = sha256_file(file);
}

const ledgerDir = path.join(ROOT, 'ledger');
if (!fs.existsSync(ledgerDir)) fs.mkdirSync(ledgerDir);

fs.writeFileSync(path.join(ledgerDir, 'sot.json'), JSON.stringify(out, null, 2) + '\n', 'utf-8');

const checksumPath = path.join(ledgerDir, 'checksum.asc');
let meta = { ...CHECKSUM_DEFAULTS };

if (fs.existsSync(checksumPath)) {
    const lines = fs.readFileSync(checksumPath, 'utf-8').split('\n');
    for (const line of lines) {
        if (line.startsWith('version:')) meta.version = line.split(':')[1].trim();
        else if (line.startsWith('revision:')) meta.revision = line.split(':')[1].trim();
        else if (line.startsWith('algorithm:')) meta.algorithm = line.split(':')[1].trim();
    }
}

const d = new Date();
const updated = d.toISOString().split('T')[0];

let ascLines = [
    "-----BEGIN ISKRA CHECKSUM-----",
    `version: ${meta.version}`,
    `revision: ${meta.revision}`,
    `updated: ${updated}`,
    `algorithm: ${meta.algorithm}`,
    "",
    "# path  sha256"
];

const sortedKeys = Object.keys(out.sha256).sort();
for (const key of sortedKeys) {
    ascLines.push(`${key}  ${out.sha256[key]}`);
}
ascLines.push("-----END ISKRA CHECKSUM-----");

fs.writeFileSync(checksumPath, ascLines.join('\n') + '\n', 'utf-8');
console.log(`Updated ledger/sot.json with ${Object.keys(out.sha256).length} entries`);
